### ISSUE TITLE
Match start and end of string

### ISSUE THREAD

**Issue by martinflack** [telekons/one-more-re-nightmare#13]
Great project. Is it intentional that a caret `^` does not match the start of the string and a dollar `$` does not match the end? Wikipedia seems to think those are included in POSIX regex.

**Comment by no-defun-allowed** [comment]
It's currently only intentional for as long as I'm still musing on the nicest way to represent it in a DFA. Gilbert told me it's possible to handle arbitrary lookahead and lookbehind with derivatives, but I still don't get his description. Or I could generalise derivatives to consume boundary information as well as characters.

**Comment by madnificent** [comment]
Is something like this on the virtual in-your-head roadmap?  I want to benchmark OMRN for matching of terminals of an LL(1) language and therefore only care about the longest match from START.

**Comment by no-defun-allowed** [comment]
It is, but I'm not sure of just what to do. At the moment it's possible I could have a function that just scans once from the start, but the issue is about handling ^ and $ which is more general, and there are a couple of ways to do them.

- The DFA and derivatives could be augmented to check conditions like start, end, start-of-word, end-of-word, etc.
- More general lookahead/behind could be implemented, if very careful.

A way to match just at the start wouldn't be too hard still. Rather than wrapping every RE in `grep` as we do, just not wrapping suffices to get a machine that scans only from the start.

**Comment by madnificent** [comment]
I don't know my way through the code well enough to give this a stab.  Could you point me to the right lines to edit or some branch so I could run an initial vague benchmark for my case?

**Comment by no-defun-allowed** [comment]
Here is a huge kludge to make OMRN generate code to only scan at the start of string. Note that it makes all regular expressions only match at the start of the string; I'd need to make a separate caches for lexing(?) and grepping functions which requires more changes. Also please pull from master as I needed to not hard-code one thing that should be a use of the `(restart)` local macro, for this to work.

```lisp
(in-package :one-more-re-nightmare)

(defclass scan-from-start ()
  ())

;;; Copied verbatim from the START-CODE method for SCAN-EVERYTHING. Oh no 
(defmethod start-code ((strategy scan-from-start) states)
  (destructuring-bind (state) states
    (let ((expression (state-expression state)))
      (cond
        ((state-never-succeeds-p state)
         ;; Just return immediately if we're told to match nothing.
         `(start (return)))
        ((re-empty-p expression)
         ;; Succeed for every character?
         `(start
           (cond
             ((> position end)
              (return))
             (t
              (incf position)
              ,(setf-from-assignments (state-exit-effects state))
              (win ,@(win-locations (state-exit-map state)))))))
        (t
         `(start
           (setf position start)
           (go ,(find-state-name state :bounds-check))))))))

(defmethod initial-states ((strategy scan-from-start) expression)
  ;; Don't do parallel GREP scanning stuff.
  (list (alpha (add-tags expression) (empty-set))))

(defmethod macros-for-strategy append ((strategy scan-from-start))
  '((restart ()
     ;; Don't actually restart the DFA.
     '(return))))

(defun make-default-strategy (layout expression)
  (declare (ignore layout expression))
  (make-instance (dynamic-mixins:mix 'scan-from-start 'call-continuation)))
```

After which:

```lisp
CL-USER> (one-more-re-nightmare:first-match "f" "food")
#(0 1)
CL-USER> (one-more-re-nightmare:first-match "f" "oodf")
NIL
```

**Comment by amno1** [comment]
A question: have you made your mind how you want this? What be needed to get omre to use this syntax:
```
ELISP> (string-match "bar$" "foobar")
3
 (#o3, #x3, ?\C-c)
ELISP> (string-match "bar$" "foobarbaz")
nil

ELISP> (string-match "^bar" "foobar")
nil

ELISP> (string-match "^bar" "barbaz")
0
 (#o0, #x0, ?\C-@)
```

Another unrelated question: is there any idea to try to accelerate omre with hyperscan/vectorscan under the hood? It would be better to implement simd acceleration directly via sb-simd or something like that as you have already done with AVX2, but I guess it is a lot of the work they have done with hyperscan :). It would probably benefit only searches in long texts or with lots of groups or alternate regex parts to search in parallel.

I need ere regexes, but I am not sure if I should write a front-end to cl-ppcre (if there already isn't one), or use omre, which I would prefer, since it seems to be faster, according to what you write in the readme, or if I would just compile a .so from gnulib and use their regex engine via cffi.

**Comment by no-defun-allowed** [comment]
> A question: have you made your mind how you want this?

I honestly haven't thought about it since, or if I have, I've forgotten about it. Sorry. All I remember is that `^` at the very start and `$` at the very end are pretty straightforward, but you *could* put those anywhere in the regex if you so wanted, and it makes for silly derivative rules.

> is there any idea to try to accelerate omre with hyperscan/vectorscan under the hood?

I don't intend to, nor am I too sure how possible it'd be. one-more-re-nightmare supports submatching and complement and intersection of regular expressions too. That doesn't entirely rule things out, since you could use Hyperscan/the like when a regex doesn't use those features, and fall back to vanilla one-more-re-nightmare when a regex does; but I'm not sure if that fits in this library neatly. My hunch is that Hyperscan does PCRE2 semantics whereas one-more-re-nightmare does POSIX semantics, [so the two will match differently](https://applied-langua.ge/projects/one-more-re-nightmare/Differences_between_POSIX_and_PCRE_regexen.html), and so it wouldn't be a drop-in replacement anyway.
